In [1]:
using Singular

In [2]:
# Ejercicio 2.4.4 UAG D.A. J.L. D.O. ordering = :degrevlex
R, (x, y, z) = PolynomialRing(QQ, ["x","y","z"],ordering = :degrevlex)
f1 = x^2 - 2 * x * z + 5
f2 = x * y^2 + y * z + 1
f3 = 3 * y^2 - 8 * x * z
I = Ideal(R,[f1,f2,f3])

Singular ideal over Singular Polynomial Ring (QQ),(x,y,z),(dp(3),C) with generators (x^2 - 2*x*z + 5, x*y^2 + y*z + 1, 3*y^2 - 8*x*z)

In [3]:
# Calcular la base de Groebner del ideal mediante  el FGLM algorithm
@time J = std(I)

  0.018300 seconds (27.19 k allocations: 1.384 MiB, 77.52% compilation time)


Singular ideal over Singular Polynomial Ring (QQ),(x,y,z),(dp(3),C) with generators (3*y^2 - 8*x*z, x^2 - 2*x*z + 5, 160*z^3 - 160*x*z + 415*y*z + 12*x - 30*y - 224*z + 15, 240*y*z^2 - 9*x*y + 1600*x*z + 18*y*z + 120*z^2 - 120*x + 240*z, 16*x*z^2 + 3*y*z - 40*z + 3, 3*x*y*z - 6*y*z^2 - 40*x*z + 3*x - 6*z)

Singular ideal over Singular Polynomial Ring (QQ),(x,y,z),(dp(3),C) with generators (3*y^2 - 8*x*z, x^2 - 2*x*z + 5, 160*z^3 - 160*x*z + 415*y*z + 12*x - 30*y - 224*z + 15, 240*y*z^2 - 9*x*y + 1600*x*z + 18*y*z + 120*z^2 - 120*x + 240*z, 16*x*z^2 + 3*y*z - 40*z + 3, 40*x*y*z - 3*x*y + 6*y*z + 40*z^2)

In [4]:
for i in gens(J)
    println(i)
end

3*y^2 - 8*x*z
x^2 - 2*x*z + 5
160*z^3 - 160*x*z + 415*y*z + 12*x - 30*y - 224*z + 15
240*y*z^2 - 9*x*y + 1600*x*z + 18*y*z + 120*z^2 - 120*x + 240*z
16*x*z^2 + 3*y*z - 40*z + 3
3*x*y*z - 6*y*z^2 - 40*x*z + 3*x - 6*z


In [5]:
@time KB = kbase(J)

  0.011912 seconds (15.70 k allocations: 827.848 KiB, 91.61% compilation time)


Singular ideal over Singular Polynomial Ring (QQ),(x,y,z),(dp(3),C) with generators (z^2, y*z, x*z, z, x*y, y, x, 1)

In [6]:
for i in gens(KB)
    println(i)
end

z^2
y*z
x*z
z
x*y
y
x
1


In [7]:
Matrix(I)

[x^2 - 2*x*z + 5, x*y^2 + y*z + 1, 3*y^2 - 8*x*z]

In [8]:
Matrix(J)

[3*y^2 - 8*x*z, x^2 - 2*x*z + 5, 160*z^3 - 160*x*z + 415*y*z + 12*x - 30*y - 224*z + 15, 240*y*z^2 - 9*x*y + 1600*x*z + 18*y*z + 120*z^2 - 120*x + 240*z, 16*x*z^2 + 3*y*z - 40*z + 3, 3*x*y*z - 6*y*z^2 - 40*x*z + 3*x - 6*z]

In [9]:
KBx = x*KB

Singular ideal over Singular Polynomial Ring (QQ),(x,y,z),(dp(3),C) with generators (x*z^2, x*y*z, x^2*z, x*z, x^2*y, x*y, x^2, x)

In [10]:
fk = x*z^2
fj = 16*x*z^2 + 3*y*z - 40*z + 3


16*x*z^2 + 3*y*z - 40*z + 3

In [11]:
for c in coefficients(fj)
    println(c)
 end

16
3
-40
3


In [12]:
for c in terms(fj)
    println(c)
 end

16*x*z^2


3*y*z
-40*z
3


In [13]:
for c in collect(monomials(fj))
    
    if c == fk
    println(c)
    end
 end

x*z^2


In [22]:
collect(monomials(fj))[1]==fk

true

In [15]:
monos = collect(monomials(fj))
coeffs = collect(coefficients(fj))
for i in 1:length(monos)
    m = monos[i]
    c = coeffs[i]
    println(c*m)
 end

16*x*z^2
3*y*z
-40*z
3


In [16]:
monos = collect(monomials(fj))
coeffs = collect(coefficients(fj))
if fk in monos
    println(true)
end

true


In [17]:
monos = collect(monomials(fj))
coeffs = collect(coefficients(fj))
arr = []
for i in 1:length(monos)
    m = monos[i]
    c = coeffs[i]
    if m == fk
        
    end
    println(c*m)
 end

16*x*z^2
3*y*z
-40*z
3


In [18]:
jota = collect(gens(J))
ka = collect(gens(KB))
kax = collect(gens(KB*x))

8-element Vector{spoly{n_Q}}:
 x*z^2
 x*y*z
 x^2*z
 x*z
 x^2*y
 x*y
 x^2
 x

In [19]:
function funcionPrueba(variable)
    Mx=[]
    for k in ka
        #if !(variable*k in ka)
            for j in jota 
                monos = collect(monomials(j))
                coeffs = collect(coefficients(j))
                arr=[]
                a = 1
                if variable*k in monos
                    for k2 in ka 
                        if k2 in monos || k==k2
                            falta = true
                            for i in 1:length(monos)
                                
                                if k2 == monos[i]
                                    push!(arr,-parse(Float64, string(coeffs[i])))
                                    falta = false
                                end
                                if k==k2 && variable*k2 == monos[i]
                                    a = parse(Float64, string(coeffs[i]))
                                end
                            end
                            if falta
                                push!(arr,0)
                            end
                        else
                            push!(arr,0)
                        end
                    end
                    arr2 = [ar/a for ar in arr]
                    push!(Mx,arr2)
                    break
                end
            end
        #else
            
        #end 
    end
    return Mx
end

funcionPrueba (generic function with 1 method)

In [20]:
funcionPrueba()

MethodError: MethodError: no method matching funcionPrueba()
Closest candidates are:
  funcionPrueba(!Matched::Any) at ~/ProyectoGrado/Nuevo.ipynb:1